Adding info on setting Kuryr configuration options (#27508)
openshift/openshift-docs: modules/installation-osp-kuryr-settings.adoc

> // Module included in the following assemblies:

Reviewer: Our modular docs system likes breaking things down as much as possible. It would still render the same in the end.

//
// * installing/installing_openstack/installing-openstack-installer-kuryr.adoc

[id="installation-osp-kuryr-settings_{context}"]
= Adjusting Kuryr settings in {rh-openstack} deployments

If you are deploying on {rh-openstack} with `Kuryr` there are several options
available for tuning of how Kuryr will behave in the cluster. This will allow
you to find balance between OpenStack Neutron ports usage and how quickly Kuryr

> is able to handle new pods being created.

Reviewer: Ah! Whoops.

You can adjust Kuryr settings before the installation as well as modify them
later.

.Port pools

> The Kuryr settings available for adjusting are related to *port pools* feature.
> The goal of that function is to minimize interactions with OpenStack Neutron
> in order to speed up creation of the new pods. Without port pools Kuryr would
> need to request a Neutron port creation each time a new Pod is spawned and
> port deletion when Pod gets removed. Those are relatively expensive operations,
> and the idea of port pools is to do them asynchronously whenever possible.

Reviewer: ```suggestion
A {rh-openstack} ports pool maintains a number of ports that can be used as needed for pod creation. 

Keeping ports on standby minimizes pod creation time. Without ports pools, Kuryr must request that a port be created or deleted every time a pod is created or deleted. 
```

> The feature makes Kuryr keep a _pool_ of Neutron ports ready to be connected to
> Pod on creation. Neutron ports used by Kuryr are created in a subnet and
> attached as subports to a trunk port - the OpenShift node port. As Kuryr
> keeps each namespace in a separate subnet, it effectively means that a
> separate port pool is kept for each namespace-worker pair.

Reviewer: ```suggestion
The Neutron ports that Kuryr users are created in [a common || their own] subnet and are attached as subports to a trunk port: the {product-title} cluster node port. 

Because Kuryr keeps each namespace in a separate subnet, a separate ports pool is maintained for each namespace-worker pair. 
```

Fill in the brackets, there. :)

> Options below are available to adjust port pools behavior:

Reviewer: ```suggestion
You can set the following parameters in the `install-config.yaml` file to configure ports pool behavior:
```

Author: It's not really set in install-config.yaml, I'll rephrase this.

> * `enablePortPoolsPrepopulation` allows you to enable pool prepopulation,

Reviewer: These properties should be added to the main install-config.yaml properties modules. They can be described in brief here, though.

Author: Yup, this isn't going into install-config.yaml, that's to be modified on the generated manifests.

  which will make Kuryr add ports to the pool when it is created, i.e. when
  new host is added or new namespace created. It is disabled by default,
  meaning that pool is populated with ports when first Pod matching the pool
  gets created.
* `poolMinPorts` is the minimum number of free ports that should be kept in
  the pool. If the number of free ports is lower new Neutron ports get created
  for the pool. Defaults to 1.
* `poolMaxPorts` is the maximum number of free ports kept in the pool. When
  the number is higher additional Neutron ports will get deleted. `0` disables
  that upper bound and is the default setting. If your OpenStack port quota is
  low, or you have a limited number of IP addresses on the pod network you
  should consider setting this option to some value to make sure ports are
  deleted when not needed.
* `poolBatchPorts` controls the number of Neutron ports getting created at
  once when there's need. The default is `3`.

> .Prerequisites

Reviewer: This is where a procedure module would start.

* Create the `install-config.yaml` file and complete any modifications to it.

.Procedure

> . Use the following command to create manifests:

Reviewer: ```suggestion
. From a command line, create the manifest files:
```

+
[source,terminal]
----
$ ./openshift-install create manifests --dir=<installation_directory> <1>
----
<1> For `<installation_directory>`, specify the name of the directory that
contains the `install-config.yaml` file for your cluster.

. Create a file that is named `cluster-network-03-config.yml` in the
`<installation_directory>/manifests/` directory:
+
[source,terminal]
----
$ touch <installation_directory>/manifests/cluster-network-03-config.yml <1>
----
<1> For `<installation_directory>`, specify the directory name that contains the
`manifests/` directory for your cluster.
+
After creating the file, several network configuration files are in the
`manifests/` directory, as shown:
+
[source,terminal]
----
$ ls <installation_directory>/manifests/cluster-network-*
----
+
.Example output
[source,terminal]
----
cluster-network-01-crd.yml
cluster-network-02-config.yml
cluster-network-03-config.yml
----

> . Open the `cluster-network-03-config.yml` file in an editor and enter a CR that

Reviewer: Will need to work through this a bit after other parts of the PR are resolved.

Author: Just a note - this is basically a customized copy of this module: https://github.com/openshift/openshift-docs/blob/18dcefa1bb6178d6aeaf806af0a7f1b9618aebb5/modules/nw-aws-nlb-new-cluster.adoc.

one given by OpenShift and second for VRRP connections. As the first one is
managed by OpenShift's and second by Neutron's IPAMs, those need to come from
different pools. Therefore, `openStackServiceNetwork` needs to be at least
twice the size of `serviceNetwork`, and whole `serviceNetwork` must be
overlapping with `openStackServiceNetwork`. cluster-network-operator will then
make sure VRRP IPs are taken from the ranges inside `openStackServiceNetwork`
that are not overlapping with `serviceNetwork`, effectivly preventing
conflicts. If not set, cluster-network-operator will use `serviceNetwork`
expanded by decrementing the prefix size by 1.
+
The CNO provides default values for the parameters in the CR, so you must
specify only the parameters that you want to change.

. Save the `cluster-network-03-config.yml` file and quit the text editor.
. Optional: Back up the `manifests/cluster-network-03-config.yml` file. The
installation program deletes the `manifests/` directory when creating the
cluster.
+
. Proceed with the rest of installation normally:
+
[source,terminal]
----
$ ./openshift-install create cluster --dir=<installation_directory>
----

> .Modifying Kuryr options after the installation

Reviewer: This will end up being a separate module, too. We can add it to a post-installation tasks assembly.
